kube-runtime/src/reflector/shared_store.rs

```diff
@@ -0,0 +1,401 @@
+where
+    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
+    K::DynamicType: Hash + Eq,
+    W: CreateWatcher<K>,
+{
+    watcher_provider: W,
+    reflectors: HashMap<ScopedListParams, (Store<K>, BoxStreamSubscribe<K>)>,
+}
+
+type ScopedListParams = (Option<String>, ListParams);
+type BoxStreamSubscribe<K> = StreamSubscribe<BoxStream<'static, watcher::Result<Event<K>>>>;
+
+impl<K> SharedStore<K>
+where
+    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send + Sync,
+    K::DynamicType: Default + Eq + Hash + Clone,
+{
+    pub fn new(client: Client) -> Self {
+        Self {
+            watcher_provider: WatcherFactory::new(client),
+            reflectors: HashMap::new(),
+        }
+    }
+}
+
+impl<K, W> SharedStore<K, W>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+    K::DynamicType: Default + Eq + Hash + Clone,
+    W: CreateWatcher<K> + 'static,
+{
+    pub fn run(self) -> impl Stream<Item = Arc<watcher::Result<Event<K>>>> {
+        stream::select_all(self.reflectors.into_iter().map(|(_, (_, reflector))| reflector))
+    }
+
+    pub fn namespaced(&mut self, namespace: &str, list_params: ListParams) -> Store<K>
+    where
+        K: Resource<Scope = NamespaceResourceScope>,
+    {
+        self.reflectors
+            .entry((Some(namespace.to_string()), list_params.clone()))
+            .or_insert_with(|| setup_reflector(self.watcher_provider.namespaced(namespace, list_params)))
+            .0
+            .clone()
+    }
+
+    pub fn all(&mut self, list_params: ListParams) -> Store<K> {
+        self.reflectors
+            .entry((None, list_params.clone()))
+            .or_insert_with(|| setup_reflector(self.watcher_provider.all(list_params)))
+            .0
+            .clone()
+    }
+
+    pub fn subscribe_namespaced(
+        &mut self,
+        namespace: &str,
+        list_params: ListParams,
+    ) -> impl Stream<Item = Result<Arc<watcher::Result<Event<K>>>, stream_subscribe::Error>>
+    where
+        K: Resource<Scope = NamespaceResourceScope>,
+    {
+        self.reflectors
+            .entry((Some(namespace.to_string()), list_params.clone()))
+            .or_insert_with(|| setup_reflector(self.watcher_provider.namespaced(namespace, list_params)))
+            .1
+            .subscribe()
+    }
+
+    pub fn subscribe_all(
+        &mut self,
+        list_params: ListParams,
+    ) -> impl Stream<Item = Result<Arc<watcher::Result<Event<K>>>, stream_subscribe::Error>>
+    where
+        K: Resource<Scope = NamespaceResourceScope>,
+    {
+        self.reflectors
+            .entry((None, list_params.clone()))
+            .or_insert_with(|| setup_reflector(self.watcher_provider.all(list_params)))
+            .1
+            .subscribe()
+    }
+}
+
+fn setup_reflector<K>(
+    watcher: BoxStream<'static, watcher::Result<Event<K>>>,
+) -> (Store<K>, BoxStreamSubscribe<K>)
+where
+    K: Resource + Clone + Send + Sync,
```

Comment:
Where does the `Sync` requirement come from? It's propagated everywhere here, and neither `watcher` nor `reflector` should have more than `Send` required AFAIKT.